### PR TITLE
Add support for testing in-app messages in SwiftUI previews

### PR DESCRIPTION
**Pull request by christibbs**
In core FIAM SDK: Opens up initializers for image subclasses
- See changes in `FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h`

In core FIAM UI test project, refactor to use open message initializers
- See changes in test files
`FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/FiamDisplaySwiftExample/BannerMessageViewController.swift`
`FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/FiamDisplaySwiftExample/ImageOnlyMessageViewController.swift`
`FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/FiamDisplaySwiftExample/ModalMessageViewController.swift`
`FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/FiamDisplaySwiftExample/CardMessageViewController.swift`

In FIAM SwiftUI SDK: Adds preview helper methods to create canned in-app messages
- See changes in new file `FirebaseInAppMessaging/Swift/Source/SwiftUIPreviewHelpers.swift`

In FIAM SwiftUI integration test project: Updates project to use SwiftUI preview helpers
- See changes in new file `FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Shared/ModalInAppMessageView.swift`

### CODE REVIEW

File: FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingDataClasses.m [hunk 1]
```diff
@@ -149,53 +150,54 @@ - (instancetype)initWithMessageID:(NSString *)messageID
...
                        actionURL:actionURL
-                         appData:nil];
+                         appData:appData];
 }
 
 @end
 
-@implementation FIRInAppMessagingModalDisplay
-
+@implementation FIRInAppMessagingBannerDisplay
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingDataClasses.m:200]: Same here re:intentional vs bad git diff?

File: FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingPrivate.h [hunk 2]
```diff
@@ -91,7 +68,7 @@ NS_ASSUME_NONNULL_BEGIN
               renderAsTestMessage:(BOOL)renderAsTestMessage
                       triggerType:(FIRInAppMessagingDisplayTriggerType)triggerType
                         titleText:(NSString *)title
-                         bodyText:(NSString *)bodyText
+                         bodyText:(nullable NSString *)bodyText
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingPrivate.h:71]: Is this an expected change?

File: FirebaseInAppMessaging/Swift/Source/SwiftUIPreviewHelpers.swift [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+import FirebaseInAppMessaging
+
+public struct InAppMessagingPreviewHelpers {
+  public static func cardMessage(campaignName: String = "Card message campaign",
+                                 title: String = "Title for modal message",
+                                 body: String? = "Body for modal message",
+                                 textColor: UIColor = UIColor.label,
+                                 backgroundColor: UIColor = UIColor.systemBackground,
+                                 portraitImage: UIImage = UIImage(systemName: "message")!,
+                                 landscapeImage: UIImage? = UIImage(systemName: "message.fill"),
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Source/SwiftUIPreviewHelpers.swift:24]: I wonder if we should use rectangles here: `rectangle.portrait` and `rectangle` (optionally adding `.fill` to either)

That way it shows the frame appropriately. WDYT?

File: FirebaseInAppMessaging/Swift/Source/SwiftUIPreviewHelpers.swift [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      appData: appData
+    )
+  }
+
+  public static func modalMessage(campaignName: String = "Modal message campaign",
+                                  title: String = "Title for modal message",
+                                  body: String? = "Body for modal message",
+                                  textColor: UIColor = UIColor.black,
+                                  backgroundColor: UIColor = UIColor.white,
+                                  image: UIImage? = UIImage(systemName: "message"),
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Source/SwiftUIPreviewHelpers.swift:81]: Same here re: using rectangle

File: FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/InAppMessagingDisplay-Sample.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -417,7 +424,7 @@
 				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
-				IPHONEOS_DEPLOYMENT_TARGET = 11.4;
+				IPHONEOS_DEPLOYMENT_TARGET = 12.0;
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/InAppMessagingDisplay-Sample.xcodeproj/project.pbxproj:427]: Was this an intentional change? Just want to make sure that FIAM still works with iOS 11

File: FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingDataClasses.m [hunk 6]
```diff
@@ -103,42 +93,53 @@ - (instancetype)initWithMessageID:(NSString *)messageID
...
                 primaryActionURL:primaryActionURL
-                         appData:nil];
+              secondaryActionURL:secondaryActionURL
+                         appData:appData];
 }
 
 @end
 
-@implementation FIRInAppMessagingBannerDisplay
+@implementation FIRInAppMessagingModalDisplay
```
[the remark this answers is not in the document]
Comment (christibbs) [reply FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingDataClasses.m:141]: Bad diff.

File: FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingDataClasses.m (hunk 1 above)
Comment (christibbs) [reply FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingDataClasses.m:200]: Bad diff.

File: FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingPrivate.h (hunk 2 above)
Comment (christibbs) [reply FirebaseInAppMessaging/Sources/RenderingObjects/FIRInAppMessagingRenderingPrivate.h:71]: Yeah, I noticed there was a nullability error there. All `bodyText` params in in-app messages should be nullable.

File: FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/InAppMessagingDisplay-Sample.xcodeproj/project.pbxproj (hunk 5 above)
Comment (christibbs) [reply FirebaseInAppMessaging/Tests/Integration/DefaultUITestApp/InAppMessagingDisplay-Sample.xcodeproj/project.pbxproj]: Reverted.

**Comment by christibbs** [comment]
@ryanwilson Resolved all open items, PTAL!

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -15,8 +15,8 @@
 		25122B2125D3482F009EE4BE /* Assets.xcassets in Resources */ = {isa = PBXBuildFile; fileRef = 25122B0B25D3482F009EE4BE /* Assets.xcassets */; };
 		25122B2D25D34F10009EE4BE /* Podfile in Resources */ = {isa = PBXBuildFile; fileRef = 25122B2C25D34F10009EE4BE /* Podfile */; };
 		25122B2E25D34F10009EE4BE /* Podfile in Resources */ = {isa = PBXBuildFile; fileRef = 25122B2C25D34F10009EE4BE /* Podfile */; };
-		2578F0A825EC13C200081E42 /* GoogleService-Info.plist in Resources */ = {isa = PBXBuildFile; fileRef = 2578F0A725EC13C200081E42 /* GoogleService-Info.plist */; };
-		2578F0A925EC13C200081E42 /* GoogleService-Info.plist in Resources */ = {isa = PBXBuildFile; fileRef = 2578F0A725EC13C200081E42 /* GoogleService-Info.plist */; };
```
[the remark this answers is not in the document]
Comment (ryanwilson) [reply FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj:19]: These are still removed

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -31,7 +31,8 @@
 		25122B1A25D3482F009EE4BE /* Info.plist */ = {isa = PBXFileReference; lastKnownFileType = text.plist.xml; path = Info.plist; sourceTree = "<group>"; };
 		25122B1B25D3482F009EE4BE /* macOS.entitlements */ = {isa = PBXFileReference; lastKnownFileType = text.plist.entitlements; path = macOS.entitlements; sourceTree = "<group>"; };
 		25122B2C25D34F10009EE4BE /* Podfile */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text; path = Podfile; sourceTree = SOURCE_ROOT; xcLanguageSpecificationIdentifier = xcode.lang.ruby; };
-		2578F0A725EC13C200081E42 /* GoogleService-Info.plist */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text.plist.xml; path = "GoogleService-Info.plist"; sourceTree = "<group>"; };
+    2578F0A725EC13C200081E42 /* GoogleService-Info.plist */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text.plist.xml; path = "GoogleService-Info.plist"; sourceTree = "<group>"; };
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj:34]: Nit: spacing doesn't match the old one

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -70,9 +71,10 @@
 		25122B0825D3482E009EE4BE /* Shared */ = {
 			isa = PBXGroup;
 			children = (
-				2578F0A725EC13C200081E42 /* GoogleService-Info.plist */,
+        2578F0A725EC13C200081E42 /* GoogleService-Info.plist */,
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj:74]: Nit: spacing

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -204,7 +206,7 @@
 			files = (
 				25122B2D25D34F10009EE4BE /* Podfile in Resources */,
 				25122B2025D3482F009EE4BE /* Assets.xcassets in Resources */,
-				2578F0A825EC13C200081E42 /* GoogleService-Info.plist in Resources */,
+        2578F0A825EC13C200081E42 /* GoogleService-Info.plist in Resources */,
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj:209]: Nit: spacing

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -214,7 +216,7 @@
 			files = (
 				25122B2E25D34F10009EE4BE /* Podfile in Resources */,
 				25122B2125D3482F009EE4BE /* Assets.xcassets in Resources */,
-				2578F0A925EC13C200081E42 /* GoogleService-Info.plist in Resources */,
+        2578F0A925EC13C200081E42 /* GoogleService-Info.plist in Resources */,
```
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/FIAMSwiftUI.xcodeproj/project.pbxproj:219]: Nit: spacing

**Review by ryanwilson — APPROVED** [review]
Just a couple small things with the xcproject but otherwise LGTM. Thanks!

Indiana Jones getting his hat at the last second

**Review by ryanwilson — APPROVED** [review]

**Pull request merged**